File: app/services/order_status_service.py
"""Avance de estado de un pedido.

Valida contra la maquina de estados que cada tenant configuro (nunca contra
codigos hardcodeados) y exige el permiso que declara cada transicion. Cocina
y caja usan este mismo camino: lo unico que cambia es el permiso que pide la
transicion configurada.
"""

import uuid
from datetime import datetime, timezone

from sqlalchemy.orm import Session

from app.domain.status_machine import Status, StatusMachine, Transition, TransitionError
from app.models.order import Order
from app.models.order_status import OrderStatus
from app.repositories import delivery_repository, order_repository, order_status_repository
from app.services.payment_service import get_balance_for_order


class OrderStatusError(Exception):
    pass


def build_machine(db: Session, tenant_id: uuid.UUID) -> tuple[StatusMachine, dict[str, OrderStatus]]:
    statuses = order_status_repository.list_statuses(db, tenant_id)
    transitions = order_status_repository.list_transitions(db, tenant_id)
    machine = StatusMachine(
        statuses=[
            Status(id=str(s.id), code=s.code, category=s.category, is_initial=s.is_initial, is_final=s.is_final)
            for s in statuses
        ],
        transitions=[
            Transition(
                from_status_id=str(t.from_status_id),
                to_status_id=str(t.to_status_id),
                required_permission=t.required_permission,
            )
            for t in transitions
        ],
    )
    return machine, {str(s.id): s for s in statuses}


def allowed_next_statuses(db: Session, *, tenant_id: uuid.UUID, order: Order) -> list[OrderStatus]:
    machine, statuses_by_id = build_machine(db, tenant_id)
    return [statuses_by_id[s.id] for s in machine.allowed_from(str(order.status_id))]


def _marcar_entrega(db: Session, order: Order, categoria: str) -> None:
    """Anota cuándo salió y cuándo llegó un domicilio.

    Se guía por la categoría del estado y no por su código: un restaurante
    puede llamar 'En moto' a lo que otro llama 'En camino', y ambos son
    'in_transit'. Los pedidos sin domicilio no tienen nada que anotar.
    """
    if categoria not in ("in_transit", "completed"):
        return

    info = delivery_repository.get_for_order(db, order.id)
    if info is None:
        return

    ahora = datetime.now(timezone.utc)
    if categoria == "in_transit" and info.dispatched_at is None:
        info.dispatched_at = ahora
    elif categoria == "completed" and info.delivered_at is None:
        info.delivered_at = ahora


def advance_status(
    db: Session,
    *,
    tenant_id: uuid.UUID,
    order_id: uuid.UUID,
    to_status_id: uuid.UUID,
    permissions: list[str],
    changed_by: uuid.UUID | None,
    note: str | None = None,
) -> Order:
    order = order_repository.get_by_id(db, tenant_id, order_id)
    if order is None:
        raise OrderStatusError("Pedido no encontrado")

    machine, statuses_by_id = build_machine(db, tenant_id)
    target = statuses_by_id.get(str(to_status_id))
    if target is None:
        raise OrderStatusError("El estado no existe para este tenant")

    try:
        machine.validate(str(order.status_id), str(to_status_id), permissions)
    except TransitionError as exc:
        raise OrderStatusError(str(exc)) from exc

    # Regla de caja: un pedido no se da por completado sin estar saldado.
    # Se apoya en la categoria, no en el nombre del estado, y deja fuera
    # 'cancelled' (tambien final) porque un pedido sin pagar si se cancela.
    if target.category == "completed":
        balance = get_balance_for_order(db, order)
        if not balance.is_settled:
            raise OrderStatusError(f"El pedido no está saldado: faltan {balance.pending}")

    order.status_id = target.id
    order.updated_at = datetime.now(timezone.utc)
    _marcar_entrega(db, order, target.category)
    order_repository.add_status_history(
        db, order_id=order.id, status_id=target.id, changed_by=changed_by, note=note
    )

    db.commit()
    db.refresh(order)
    return order

File: php/src/Services/OrderStatusService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Core\Database;
use App\Domain\StatusChangeError;
use App\Domain\StatusChangeRules;
use App\Domain\StatusMachine;
use App\Domain\TransitionError;
use App\Models\Order;
use App\Models\OrderStatusRow;
use App\Repositories\OrderRepository;
use App\Repositories\OrderStatusRepository;

/**
 * Avance de estado de un pedido.
 *
 * Valida contra la maquina de estados que cada tenant configuro (nunca contra
 * codigos hardcodeados) y exige el permiso que declara cada transicion. Cocina
 * y caja usan este mismo camino: lo unico que cambia es el permiso que pide la
 * transicion configurada.
 */
final class OrderStatusService
{
    /** @return array{0: StatusMachine, 1: array<string, OrderStatusRow>} */
    public static function buildMachine(string $tenantId): array
    {
        $repo = new OrderStatusRepository(Database::app());
        $statuses = $repo->listStatuses($tenantId);

        $machine = new StatusMachine(
            array_map(static fn (OrderStatusRow $s) => $s->toDomain(), $statuses),
            $repo->listTransitions($tenantId),
        );

        $byId = [];
        foreach ($statuses as $status) {
            $byId[$status->id] = $status;
        }
        return [$machine, $byId];
    }

    /** @return OrderStatusRow[] */
    public static function allowedNextStatuses(string $tenantId, Order $order): array
    {
        [$machine, $byId] = self::buildMachine($tenantId);
        return array_map(
            static fn ($s) => $byId[$s->id],
            $machine->allowedFrom($order->statusId),
        );
    }

    /** @param string[] $permissions */
    public static function advanceStatus(
        string $tenantId,
        string $orderId,
        string $toStatusId,
        array $permissions,
        ?string $changedBy,
        ?string $note = null,
    ): Order {
        $orders = new OrderRepository(Database::app());

        // Con bloqueo: dos cajeros que avancen el mismo pedido a la vez ya no
        // se pisan, el segundo espera y valida contra el estado que quedo.
        $order = $orders->getByIdForUpdate($tenantId, $orderId);
        if ($order === null) {
            throw new OrderStatusError('Pedido no encontrado');
        }

        [$machine, $byId] = self::buildMachine($tenantId);
        $target = $byId[$toStatusId] ?? null;
        if ($target === null) {
            throw new OrderStatusError('El estado no existe para este tenant');
        }

        try {
            $machine->validate($order->statusId, $toStatusId, $permissions);
        } catch (TransitionError $e) {
            throw new OrderStatusError($e->getMessage());
        }

        // Lo que el dinero del pedido permite, por encima de lo que la
        // maquina de estados del tenant configuro: completar exige estar
        // saldado, anular exige no tener plata encima y decir por que.
        // Ambas reglas son de la plataforma, no de cada restaurante, y viven
        // juntas en Domain\StatusChangeRules.
        try {
            StatusChangeRules::ensureCanEnter(
                $target->category,
                PaymentService::getBalanceForOrder($order),
                $note,
            );
        } catch (StatusChangeError $e) {
            throw new OrderStatusError($e->getMessage());
        }

        $orders->setStatus($order->id, $target->id);
        $orders->addStatusHistory($order->id, $target->id, $changedBy, $note);

        // Modulo 6: si el pedido es un domicilio, aqui queda la hora de salida
        // o de entrega. Va despues del cambio de estado y no antes para que un
        // rechazo de la maquina de estados no deje una hora sellada de un
        // avance que nunca ocurrio.
        DeliveryService::stampByCategory($order->id, $target->category);

        return $orders->getById($tenantId, $order->id);
    }
}
